File: algorithms/C/Complete Binary Tree Inserter/Complete Binary Tree Inserter.py
class CBTInserter:

    def __init__(self, root: Optional[TreeNode]):
        self.root = root
        self.queue = Queue()
        self.queue.put(self.root)
        self.parent_of_last_inserted = None
        
    def insert(self, val: int) -> int:
        if self.parent_of_last_inserted is not None and self.parent_of_last_inserted.right is None:
            self.parent_of_last_inserted.right = TreeNode(val)
            self.queue.put(self.parent_of_last_inserted.right)
            return self.parent_of_last_inserted.val
        
        while not self.queue.empty():
            node = self.queue.get()
            if node.left is None:
                node.left = TreeNode(val)
                self.queue.put(node.left)
                self.parent_of_last_inserted = node
                return node.val
            else:
                self.queue.put(node.left)
            if node.right is None:
                node.right = TreeNode(val)
                self.queue.put(node.right)
                self.parent_of_last_inserted = node
                return node.val
            else:
                self.queue.put(node.right)
        

    def get_root(self) -> Optional[TreeNode]:
        return self.root

File: algorithms/C/Complete Binary Tree Inserter/Complete Binary Tree Inserter.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
var CBTInserter = function(root) {
    this.root = root;
};

/** 
 * @param {number} val
 * @return {number}
 */
CBTInserter.prototype.insert = function(val) {
    let cur = this.root;
    let queue = [cur];
    while(queue.length){
        let updatedQueue = [];
        while(queue.length){
            let node = queue.shift();
            if(!node.left) {
                node.left = new TreeNode(val);
                return node.val;
            }
            else updatedQueue.push(node.left);
            if(!node.right) {
                node.right = new TreeNode(val);
                return node.val;
            }
            else updatedQueue.push(node.right);
        }
        queue = updatedQueue ;
    }
    
};

/**
 * @return {TreeNode}
 */
CBTInserter.prototype.get_root = function() {
    return this.root;
};

/** 
 * Your CBTInserter object will be instantiated and called as such:
 * var obj = new CBTInserter(root)
 * var param_1 = obj.insert(val)
 * var param_2 = obj.get_root()
 */
